File: banki/grading.py
"""Verse dictation grading with punctuation tolerance and fuzzy matching."""
import re
import unicodedata

PUNCT_PATTERN = re.compile(r"[　\.,;:!?\"'`~\-—–…()\[\]{}<>/\\|·•‘’“”]")


def _strip_punct(text: str) -> str:
    return PUNCT_PATTERN.sub("", text)


def _normalize(text: str) -> str:
    text = unicodedata.normalize("NFC", text or "")
    text = _strip_punct(text)
    text = text.replace(" ", "").replace("　", "")
    return text.lower()


def _split_words(text: str) -> list[str]:
    text = unicodedata.normalize("NFC", text or "")
    text = _strip_punct(text)
    return [w for w in text.split() if w]


def _levenshtein(a: str, b: str) -> int:
    if not a:
        return len(b)
    if not b:
        return len(a)
    if a == b:
        return 0
    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a, 1):
        curr = [i] + [0] * len(b)
        for j, cb in enumerate(b, 1):
            cost = 0 if ca == cb else 1
            curr[j] = min(curr[j - 1] + 1, prev[j] + 1, prev[j - 1] + cost)
        prev = curr
    return prev[-1]


def _similarity(a: str, b: str) -> float:
    a_n, b_n = _normalize(a), _normalize(b)
    if not a_n and not b_n:
        return 1.0
    if not a_n or not b_n:
        return 0.0
    dist = _levenshtein(a_n, b_n)
    return 1 - dist / max(len(a_n), len(b_n))


def compute_word_match(user_text: str, answer_text: str, fuzzy_threshold: float = 0.6) -> dict:
    """Compare user input with answer text word-by-word.

    A word is "matched" if it's an exact normalized match, or a fuzzy match
    above `fuzzy_threshold`. Fuzzy matches are marked as partial.
    """
    answer_words = _split_words(answer_text)
    user_words = _split_words(user_text)

    if not answer_words:
        return {
            "score": 100, "total_words": 0, "matched_words": 0,
            "partial_words": 0,
            "answer_words": [], "user_words": [], "word_results": [],
        }

    matched = 0
    partial = 0
    word_results = []

    for i, aw in enumerate(answer_words):
        if i < len(user_words):
            uw = user_words[i]
            sim = _similarity(uw, aw)
            if sim == 1.0:
                matched += 1
                word_results.append({"answer": aw, "user": uw, "match": "full", "similarity": sim})
            elif sim >= fuzzy_threshold:
                partial += 1
                word_results.append({"answer": aw, "user": uw, "match": "partial", "similarity": sim})
            else:
                word_results.append({"answer": aw, "user": uw, "match": "miss", "similarity": sim})
        else:
            word_results.append({"answer": aw, "user": "", "match": "missing", "similarity": 0.0})

    for i in range(len(answer_words), len(user_words)):
        word_results.append({"answer": "", "user": user_words[i], "match": "extra", "similarity": 0.0})

    score_raw = (matched + partial * 0.6) / len(answer_words) * 100
    score = max(0, min(100, round(score_raw)))

    return {
        "score": score,
        "total_words": len(answer_words),
        "matched_words": matched,
        "partial_words": partial,
        "answer_words": answer_words,
        "user_words": user_words,
        "word_results": word_results,
    }


def render_word_comparison_html(result: dict) -> str:
    """Render word-by-word comparison with color coding."""
    html_parts = []
    for wr in result["word_results"]:
        m = wr["match"]
        if m == "full":
            html_parts.append(
                f'<span class="w-full">{wr["answer"]}</span>'
            )
        elif m == "partial":
            html_parts.append(
                f'<span class="w-partial" title="유사도 {int(wr["similarity"]*100)}%">'
                f'{wr["user"]} → {wr["answer"]}</span>'
            )
        elif m == "missing":
            html_parts.append(
                f'<span class="w-missing">{wr["answer"]}</span>'
            )
        elif m == "extra":
            html_parts.append(
                f'<span class="w-extra"><s>{wr["user"]}</s></span>'
            )
        else:
            html_parts.append(
                f'<span class="w-miss"><s>{wr["user"]}</s> → {wr["answer"]}</span>'
            )
    return " ".join(html_parts)

File: web/js/lib/grading.js
// Verse dictation grading with punctuation tolerance and fuzzy matching.
// Ported from banki/grading.py; data shapes keep the original snake_case keys
// so the two implementations can be compared directly by the fixture tests.

import { pyRound, escapeHtml } from './util.js';

const PUNCT = /[　.,;:!?"'`~\-—–…()[\]{}<>\/\\|·•‘’“”]/g;

function stripPunct(text) {
  return text.replace(PUNCT, '');
}

function normalize(text) {
  let out = (text || '').normalize('NFC');
  out = stripPunct(out);
  out = out.replace(/ /g, '').replace(/　/g, '');
  return out.toLowerCase();
}

function splitWords(text) {
  const out = stripPunct((text || '').normalize('NFC'));
  return out.split(/\s+/).filter(Boolean);
}

function levenshtein(a, b) {
  if (!a) return b.length;
  if (!b) return a.length;
  if (a === b) return 0;
  let prev = Array.from({ length: b.length + 1 }, (_, i) => i);
  for (let i = 1; i <= a.length; i++) {
    const curr = new Array(b.length + 1);
    curr[0] = i;
    for (let j = 1; j <= b.length; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min(curr[j - 1] + 1, prev[j] + 1, prev[j - 1] + cost);
    }
    prev = curr;
  }
  return prev[b.length];
}

function similarity(a, b) {
  const an = normalize(a);
  const bn = normalize(b);
  if (!an && !bn) return 1.0;
  if (!an || !bn) return 0.0;
  return 1 - levenshtein(an, bn) / Math.max(an.length, bn.length);
}

/**
 * Compare user input with the answer text word by word.
 *
 * A word counts as matched when it normalizes to the same string, and as a
 * partial match when its similarity clears `fuzzyThreshold`.
 */
export function computeWordMatch(userText, answerText, fuzzyThreshold = 0.6) {
  const answerWords = splitWords(answerText);
  const userWords = splitWords(userText);

  if (!answerWords.length) {
    return {
      score: 100,
      total_words: 0,
      matched_words: 0,
      partial_words: 0,
      answer_words: [],
      user_words: [],
      word_results: [],
    };
  }

  let matched = 0;
  let partial = 0;
  const wordResults = [];

  answerWords.forEach((aw, i) => {
    if (i < userWords.length) {
      const uw = userWords[i];
      const sim = similarity(uw, aw);
      if (sim === 1.0) {
        matched++;
        wordResults.push({ answer: aw, user: uw, match: 'full', similarity: sim });
      } else if (sim >= fuzzyThreshold) {
        partial++;
        wordResults.push({ answer: aw, user: uw, match: 'partial', similarity: sim });
      } else {
        wordResults.push({ answer: aw, user: uw, match: 'miss', similarity: sim });
      }
    } else {
      wordResults.push({ answer: aw, user: '', match: 'missing', similarity: 0.0 });
    }
  });

  for (let i = answerWords.length; i < userWords.length; i++) {
    wordResults.push({ answer: '', user: userWords[i], match: 'extra', similarity: 0.0 });
  }

  const rawScore = ((matched + partial * 0.6) / answerWords.length) * 100;
  const score = Math.max(0, Math.min(100, pyRound(rawScore)));

  return {
    score,
    total_words: answerWords.length,
    matched_words: matched,
    partial_words: partial,
    answer_words: answerWords,
    user_words: userWords,
    word_results: wordResults,
  };
}

/** Word-by-word comparison markup, colour-coded by match kind. */
export function renderWordComparison(result) {
  return result.word_results
    .map((wr) => {
      const answer = escapeHtml(wr.answer);
      const user = escapeHtml(wr.user);
      switch (wr.match) {
        case 'full':
          return `<span class="w-full">${answer}</span>`;
        case 'partial':
          return `<span class="w-partial" title="유사도 ${Math.round(wr.similarity * 100)}%">${user} → ${answer}</span>`;
        case 'missing':
          return `<span class="w-missing">${answer}</span>`;
        case 'extra':
          return `<span class="w-extra"><s>${user}</s></span>`;
        default:
          return `<span class="w-miss"><s>${user}</s> → ${answer}</span>`;
      }
    })
    .join(' ');
}
